### DISCUSSION TITLE
Web3 importing no clear solution in code

### DISCUSSION THREAD

**Discussion by LadiesMan1309** [smartcontractkit/full-blockchain-solidity-course-py#1816] [Q&A] [answered]
haricharnyarlagadda@Haricharns-MBP web3_py_simple_storage % pip install w
eb3
Collecting web3
  Using cached web3-5.31.3-py3-none-any.whl (501 kB)
Collecting aiohttp<4,>=3.7.4.post0
  Using cached aiohttp-3.8.3-cp311-cp311-macosx_10_9_x86_64.whl (355 kB)
Collecting eth-abi<3.0.0,>=2.2.0
  Using cached eth_abi-2.2.0-py3-none-any.whl (28 kB)
Collecting eth-account<0.6.0,>=0.5.9
  Using cached eth_account-0.5.9-py3-none-any.whl (101 kB)
Collecting eth-hash[pycryptodome]<1.0.0,>=0.2.0
  Using cached eth_hash-0.5.1-py3-none-any.whl (9.0 kB)
Collecting eth-rlp<0.3
  Using cached eth_rlp-0.2.1-py3-none-any.whl (5.0 kB)
Collecting eth-typing<3.0.0,>=2.0.0
  Using cached eth_typing-2.3.0-py3-none-any.whl (6.2 kB)
Collecting eth-utils<2.0.0,>=1.9.5
  Using cached eth_utils-1.10.0-py3-none-any.whl (24 kB)
Collecting hexbytes<1.0.0,>=0.1.0
  Using cached hexbytes-0.3.0-py3-none-any.whl (6.4 kB)
Collecting ipfshttpclient==0.8.0a2
  Using cached ipfshttpclient-0.8.0a2-py3-none-any.whl (82 kB)
Collecting jsonschema<5,>=3.2.0
  Using cached jsonschema-4.17.3-py3-none-any.whl (90 kB)
Collecting lru-dict<2.0.0,>=1.1.6
  Using cached lru-dict-1.1.8.tar.gz (10 kB)
  Preparing metadata (setup.py) ... done
Collecting protobuf==3.19.5
  Using cached protobuf-3.19.5-py2.py3-none-any.whl (162 kB)
Requirement already satisfied: requests<3.0.0,>=2.16.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from web3) (2.28.1)
Collecting websockets<10,>=9.1
  Using cached websockets-9.1-cp311-cp311-macosx_10_9_universal2.whl
Collecting multiaddr>=0.0.7
  Using cached multiaddr-0.0.9-py2.py3-none-any.whl (16 kB)
Collecting attrs>=17.3.0
  Using cached attrs-22.2.0-py3-none-any.whl (60 kB)
Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from aiohttp<4,>=3.7.4.post0->web3) (2.1.1)
Collecting multidict<7.0,>=4.5
  Using cached multidict-6.0.4-cp311-cp311-macosx_10_9_x86_64.whl (29 kB)
Collecting async-timeout<5.0,>=4.0.0a3
  Using cached async_timeout-4.0.2-py3-none-any.whl (5.8 kB)
Collecting yarl<2.0,>=1.0
  Using cached yarl-1.8.2-cp311-cp311-macosx_10_9_x86_64.whl (59 kB)
Collecting frozenlist>=1.1.1
  Using cached frozenlist-1.3.3-cp311-cp311-macosx_10_9_x86_64.whl (35 kB)
Collecting aiosignal>=1.1.2
  Using cached aiosignal-1.3.1-py3-none-any.whl (7.6 kB)
Collecting parsimonious<0.9.0,>=0.8.0
  Using cached parsimonious-0.8.1-py3-none-any.whl
Collecting bitarray<3,>=1.2.1
  Using cached bitarray-2.6.1-cp311-cp311-macosx_10_9_x86_64.whl (105 kB)
Collecting eth-keyfile<0.6.0,>=0.5.0
  Using cached eth_keyfile-0.5.1-py3-none-any.whl (8.3 kB)
Collecting eth-keys<0.4.0,>=0.3.4
  Using cached eth_keys-0.3.4-py3-none-any.whl (21 kB)
Collecting rlp<3,>=1.0.0
  Using cached rlp-2.0.1-py2.py3-none-any.whl (20 kB)
Collecting pycryptodome<4,>=3.6.6
  Using cached pycryptodome-3.16.0-cp35-abi3-macosx_10_9_x86_64.whl (1.6 MB)
Collecting eth-utils<2.0.0,>=1.9.5
  Using cached eth_utils-1.9.5-py3-none-any.whl (23 kB)
Collecting cytoolz<1.0.0,>=0.10.1
  Using cached cytoolz-0.12.1-cp311-cp311-macosx_10_9_x86_64.whl (398 kB)
Collecting pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0
  Using cached pyrsistent-0.19.2-py3-none-any.whl (57 kB)
Requirement already satisfied: idna<4,>=2.5 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from requests<3.0.0,>=2.16.0->web3) (3.4)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from requests<3.0.0,>=2.16.0->web3) (1.26.13)
Requirement already satisfied: certifi>=2017.4.17 in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from requests<3.0.0,>=2.16.0->web3) (2022.12.7)
Collecting toolz>=0.8.0
  Using cached toolz-0.12.0-py3-none-any.whl (55 kB)
Requirement already satisfied: varint in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from multiaddr>=0.0.7->ipfshttpclient==0.8.0a2->web3) (1.0.2)
Collecting six
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting base58
  Using cached base58-2.1.1-py3-none-any.whl (5.6 kB)
Requirement already satisfied: netaddr in /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages (from multiaddr>=0.0.7->ipfshttpclient==0.8.0a2->web3) (0.8.0)
Building wheels for collected packages: lru-dict
  Building wheel for lru-dict (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [9 lines of output]
      running bdist_wheel
      running build
      running build_ext
      building 'lru' extension
      creating build
      creating build/temp.macosx-10.9-universal2-cpython-311
      clang -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -arch arm64 -arch x86_64 -g -I/Library/Frameworks/Python.framework/Versions/3.11/include/python3.11 -c lru.c -o build/temp.macosx-10.9-universal2-cpython-311/lru.o
      xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for lru-dict
  Running setup.py clean for lru-dict
Failed to build lru-dict
Installing collected packages: lru-dict, bitarray, websockets, toolz, six, pyrsistent, pycryptodome, protobuf, multidict, hexbytes, frozenlist, eth-typing, eth-hash, base58, attrs, async-timeout, yarl, parsimonious, multiaddr, jsonschema, cytoolz, aiosignal, ipfshttpclient, eth-utils, aiohttp, rlp, eth-keys, eth-abi, eth-rlp, eth-keyfile, eth-account, web3
  Running setup.py install for lru-dict ... error
  error: subprocess-exited-with-error
  
  × Running setup.py install for lru-dict did not run successfully.
  │ exit code: 1
  ╰─> [11 lines of output]
      running install
      /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_ext
      building 'lru' extension
      creating build
      creating build/temp.macosx-10.9-universal2-cpython-311
      clang -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -arch arm64 -arch x86_64 -g -I/Library/Frameworks/Python.framework/Versions/3.11/include/python3.11 -c lru.c -o build/temp.macosx-10.9-universal2-cpython-311/lru.o
      xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun
      error: command '/usr/bin/clang' failed with exit code 1
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> lru-dict

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.

**Comment by cromewar** [comment, marked as answer]
Hello @YeetBrain 
This could happen due to several reasons:
1. First of all python 3.11 is not yet fully compatible with web3 libraries so is better to use 3.8 o 3.10 tops.
2. Secondly you are mussing some programming dependencies and libraries for your Mac, so you need to install `Xcode` on your mac and also `Cython`, please check this [this](https://medium.com/@cromewar/the-mighty-cytools-error-at-web3-smart-contract-development-with-brownie-63335d50f230) link.

I hope this info might help, cheers.
